nomad/host_volume_endpoint_test.go

```diff
@@ -0,0 +1,571 @@
+			Alloc: []*structs.Allocation{alloc},
+			WriteRequest: structs.WriteRequest{
+				Region:    srv.Region(),
+				AuthToken: node1.SecretID},
+		}
+		err = msgpackrpc.CallWithCodec(codec, "Node.UpdateAlloc", nArgs, &structs.GenericResponse{})
+
+		delReq.VolumeIDs = []string{vol2ID}
+		err = msgpackrpc.CallWithCodec(codec, "HostVolume.Delete", delReq, &delResp)
+		must.NoError(t, err)
+
+		getReq.ID = vol2ID
+		err = msgpackrpc.CallWithCodec(codec, "HostVolume.Get", getReq, &getResp)
+		must.NoError(t, err)
+		must.Nil(t, getResp.Volume)
+	})
+}
+
+func TestHostVolumeEndpoint_List(t *testing.T) {
+
+	srv, rootToken, cleanupSrv := TestACLServer(t, func(c *Config) {
+		c.NumSchedulers = 0
+	})
+	t.Cleanup(cleanupSrv)
+	testutil.WaitForLeader(t, srv.RPC)
+	store := srv.fsm.State()
+	codec := rpcClient(t, srv)
+
+	index := uint64(1001)
+
+	token := mock.CreatePolicyAndToken(t, store, index, "volume-manager",
+		`namespace "apps" { capabilities = ["host-volume-register"] }
+	     node { policy = "read" }`).SecretID
+
+	index++
+	otherToken := mock.CreatePolicyAndToken(t, store, index, "other",
+		`namespace "foo" { capabilities = ["host-volume-read"] }
+         node { policy = "read" }`).SecretID
+
+	index++
+	ns1 := "apps"
+	ns2 := "system"
+	nspace1, nspace2 := mock.Namespace(), mock.Namespace()
+	nspace1.Name = ns1
+	nspace2.Name = ns2
+	must.NoError(t, store.UpsertNamespaces(index, []*structs.Namespace{nspace1, nspace2}))
+
+	nodes := []*structs.Node{
+		mock.Node(),
+		mock.Node(),
+		mock.Node(),
+	}
+	nodes[2].NodePool = "prod"
+	index++
+	must.NoError(t, store.UpsertNode(structs.MsgTypeTestSetup,
+		index, nodes[0], state.NodeUpsertWithNodePool))
+	must.NoError(t, store.UpsertNode(structs.MsgTypeTestSetup,
+		index, nodes[1], state.NodeUpsertWithNodePool))
+	must.NoError(t, store.UpsertNode(structs.MsgTypeTestSetup,
+		index, nodes[2], state.NodeUpsertWithNodePool))
+
+	vol1, vol2 := mock.HostVolume(), mock.HostVolume()
+	vol1.NodeID = nodes[0].ID
+	vol1.Name = "foobar-example"
+	vol1.Namespace = ns1
+	vol2.NodeID = nodes[1].ID
+	vol2.Name = "foobaz-example"
+	vol2.Namespace = ns1
+
+	// we need to register these rather than upsert them so we have the correct
+	// indexes for unblocking later
+	registerReq := &structs.HostVolumeRegisterRequest{
+		Volumes: []*structs.HostVolume{vol1, vol2},
+		WriteRequest: structs.WriteRequest{
+			Region:    srv.Region(),
+			Namespace: ns1,
+			AuthToken: rootToken.SecretID},
+	}
+
+	var registerResp structs.HostVolumeRegisterResponse
+	err := msgpackrpc.CallWithCodec(codec, "HostVolume.Register", registerReq, &registerResp)
+	must.NoError(t, err)
+
+	vol3, vol4 := mock.HostVolume(), mock.HostVolume()
+	vol3.NodeID = nodes[2].ID
+	vol3.NodePool = "prod"
+	vol3.Namespace = ns2
+	vol3.Name = "foobar-example"
+	vol4.Namespace = ns2
+	vol4.NodeID = nodes[1].ID
+	vol4.Name = "foobaz-example"
+
+	registerReq.Volumes = []*structs.HostVolume{vol3, vol4}
+	registerReq.Namespace = ns2
```

Comment:
I was wondering why these aren't registered with vol1&2, and I suppose it's because different `WriteRequest.Namespace`? I thought vols in separate NS could be registered at the same time... or have I got myself confused?

Reply:
> I thought vols in separate NS could be registered at the same time... or have I got myself confused?

No, you're right! That changed from when I wrote this test a few days ago, because of the bug we discovered in CSI. I can clean that up now.